Source/Core/Core/IOS/DI/DI.cpp

```diff
@@ -30,70 +33,513 @@ void DI::DoState(PointerWrap& p)
+  case DVDLowGetControlRegister:
+    INFO_LOG(IOS_DI, "DVDLowGetControlRegister: 0x%08x", DVDInterface::s_DICR.Hex);
+    return WriteIfFits(request, DVDInterface::s_DICR.Hex);
+  }
 
   // DVDInterface's ExecuteCommand handles most of the work.
   // The IOCtl callback is used to generate a reply afterwards.
-  DVDInterface::ExecuteCommand(command_0, command_1, command_2, request.buffer_out,
-                               request.buffer_out_size, true);
+  switch (request.request)
+  {
+  case DVDLowInquiry:
+    INFO_LOG(IOS_DI, "DVDLowInquiry");
+    return StartDMATransfer(0x12000000, 0, 0, 0x20, request);
+  case DVDLowReadDiskID:
+    INFO_LOG(IOS_DI, "DVDLowReadDiskID");
+    // TODO: also include the post-read second read
+    return StartDMATransfer(0xA8000040, 0, 0, 0x20, request);
+  case DVDLowRead:
+  {
+    // TODO.  Needs to include decryption.
+    u32 length = Memory::Read_U32(request.buffer_in + 4);
+    u32 position = Memory::Read_U32(request.buffer_in + 8);
+    INFO_LOG(IOS_DI, "DVDLowRead: offset 0x%08x (byte 0x%09" PRIx64 "), length 0x%x", position,
+             (u64)position << 2, length);
+    if (request.buffer_out_size < length)
+    {
+      WARN_LOG(IOS_DI,
+               "Output buffer is too small for the result of the read (%d bytes given, needed at "
+               "least %d); returning security error",
+               request.buffer_out_size, length);
+      return std::make_optional(SECURITY_ERROR);
+    }
+    m_last_length = position;  // An actual mistake in IOS
+    DVDInterface::PerformDecryptingRead(position, length, request.buffer_out, true);
+    return {};
+  }
+  case DVDLowReadDvdPhysical:
+  {
+    u8 position = Memory::Read_U8(request.buffer_in + 7);
+    INFO_LOG(IOS_DI, "DVDLowReadDvdPhysical: position 0x%02x", position);
+    return StartDMATransfer(0xAD000000 | (position << 8), 0, 0, 0x800, request);
+  }
+  case DVDLowReadDvdCopyright:
+  {
+    u8 position = Memory::Read_U8(request.buffer_in + 7);
+    INFO_LOG(IOS_DI, "DVDLowReadDvdCopyright: position 0x%02x", position);
+    return StartImmediateTransfer(0xAD010000 | (position << 8), 0, 0, request);
+  }
+  case DVDLowReadDvdDiscKey:
+  {
+    u8 position = Memory::Read_U8(request.buffer_in + 7);
+    INFO_LOG(IOS_DI, "DVDLowReadDvdDiscKey: position 0x%02x", position);
+    return StartDMATransfer(0xAD020000 | (position << 8), 0, 0, 0x800, request);
+  }
+  case DVDLowUnencryptedRead:
+  {
+    u32 length = Memory::Read_U32(request.buffer_in + 4);
+    u32 position = Memory::Read_U32(request.buffer_in + 8);
+    u32 end = position + (length >> 2);  // a 32-bit offset
+    INFO_LOG(IOS_DI, "DVDLowUnencryptedRead: offset 0x%08x (byte 0x%09" PRIx64 "), length 0x%x",
+             position, (u64)position << 2, length);
+    // (start, end) as 32-bit offsets
+    // Older IOS versions only accept the first range.  Later versions added the extra ranges to
+    // check how the drive responds to out of bounds reads (an error 001 check).
```

Comment:
Do you know which versions? You could add this to GetFeatures() in IOS/VersionInfo.cpp if you want Dolphin to handle this the way a given version of IOS would.

Reply:
I haven't fully tested all versions (I focused on the version dated Jun 3 2009 07:49:09 and included in IOS58 and IOS80), but here's what I wrote on WiiBrew:

> Versions of IOS prior to IOS30 only permitted reads in the first range[check]. Nintendo titles check on startup if they are running an IOS ≥ 30 (and ≤ 253) and if so, perform some DI checks; specifically, they attempt to read 0x20 bytes from 0x460a0000 (or from 0x7ed40000 if the byte at 0x8000319c is 0x81 — possibly related to dual-layer discs?). If this read attempt returns anything other than 2, the game will refuse to start with the message "An error has occurred. Press the Eject Button, remove the Game Disc, and turn off the power to the console. Please read the Wii Operations Manual for further instructions." If the drive error is anything other than 0x0052100 (OK/Logical block address out of range), the game will refuse to start with the message "Error #001, unauthorized device has been detected."